File: Lowest Common Ancestor(Java)/LowestCommonAncestor.java

class Node { 
    int data; 
    Node left; 
    Node right; 
   
    Node(int item) { 
        data = item; 
        left = right = null; 
    } 
} 
 
class LowestCommonAncestor { 
    Node root; 
    public static void main(String args[]) {                  
        LowestCommonAncestor tree = new LowestCommonAncestor();     
        int X, Y;                                                   
        tree.root = new Node(14);                   //               14           
        tree.root.left = new Node(9);               //              /  \              
        tree.root.right = new Node(12);             //             9    12               
        tree.root.left.left = new Node(6);          //            / \   / \               
        tree.root.left.right = new Node(8);         //           6   8 10  11                
        tree.root.right.left = new Node(10);        //              / \                
        tree.root.right.right = new Node(11);       //             4   1                
        tree.root.left.right.left = new Node(4);                    
        tree.root.left.right.right = new Node(1);                   

        X = 4; 
        Y = 1; 
        Node node = tree.lowestCommonAncestor(tree.root, X, Y); 
        System.out.println("Lowest Common Ancestor of " + X + " and " + Y + " is " + node.data + "."); 

        X = 10; 
        Y = 11;
        node = tree.lowestCommonAncestor(tree.root, X, Y); 
        System.out.println("Lowest Common Ancestor of " + X + " and " + Y + " is " + node.data + "."); 

        X = 6; 
        Y = 8; 
        node = tree.lowestCommonAncestor(tree.root, X, Y); 
        System.out.println("Lowest Common Ancestor of " + X + " and " + Y + " is " + node.data + "."); 

        X = 6; 
        Y = 12; 
        node = tree.lowestCommonAncestor(tree.root, X, Y); 
        System.out.println("Lowest Common Ancestor of " + X + " and " + Y + " is " + node.data + "."); 
    } 

    Node lowestCommonAncestor(Node node, int X, int Y) { 
        if (node == null) {
            return null; 
        }
  
        if (node.data == X || node.data == Y) {
            return node; 
        }

        Node leftLCA = lowestCommonAncestor(node.left, X, Y); 
        Node rightLCA = lowestCommonAncestor(node.right, X, Y); 
  
        if (leftLCA != null && rightLCA != null) {
            return node;
        } 

        if (leftLCA != null) {
            return leftLCA;
        }
        else return rightLCA;
    } 
} 

File: DAG-Lowest Common Ancestor(Python)/LowestCommonAncestor.py
class Node: 
	def __init__(self, data): 
		self.data = data 
		self.left = None
		self.right = None

def lowest_common_ancestor(root, x, y): 
    if root is None: 
        return None
  
    if root.data == x or root.data == y: 
        return root  

    left_side = lowest_common_ancestor(root.left, x, y)  
    right_side = lowest_common_ancestor(root.right, x, y) 

    if left_side and right_side: 
        return root  
 
    return left_side if left_side is not None else right_side

root = Node(14)                 #            14
root.left = Node(9)             #           /  \
root.right = Node(12)           #          9   12
root.left.left = Node(6)        #         / \  / \
root.left.right = Node(8)       #        6  8 10 11 
root.left.right.left = Node(4)  #          / \ 
root.left.right.right = Node(1) #         4   1
root.right.left = Node(10)
root.right.right = Node(11)

print ("Lowest Common Aancestor of 4 and 1 is", lowest_common_ancestor(root, 4, 1).data)
print ("Lowest Common Aancestor of 10 and 11 is", lowest_common_ancestor(root, 10, 11).data)
print ("Lowest Common Aancestor of 6 and 8 is", lowest_common_ancestor(root, 6, 8).data)
print ("Lowest Common Aancestor of 9 and 12 is", lowest_common_ancestor(root, 9, 12).data)